Dependency locking documentation

ljacomet opened the pull request.

melix commented on subprojects/docs/src/docs/userguide/dependencyLocking.adoc:
```diff
@@ -0,0 +1,129 @@
+
+It is currently not possible to lock the `classpath` configuration used for build plugins.
+
+=== Nebula locking plugin
```
:x: Let's remove this section. We don't mention external plugins in it, as it may vary over time.

ljacomet replied:
It was referenced from another section, so I though it was a good idea to move it here. And also a way of indicating that there is not a 100% match between the two different feature sets.

melix replied:
Right, I wasn't aware we mentioned it. I'd still be in favor of removing.

ljacomet commented:
@melix Aside from the Nebula reference, I have addressed your comments.
@eriwen Any chance you can review this?

eriwen commented on subprojects/docs/src/samples/userguide/dependencyManagement/dependencyLocking/lockingSingleConfiguration/build.gradle:
```diff
@@ -0,0 +1,26 @@
+
+//START SNIPPET locking-one
+configurations {
+    compileClasspath {
```
Also, not sure if we should use that if it's even possible.

eriwen commented on subprojects/docs/src/samples/userguide/dependencyManagement/dependencyLocking/lockingSingleConfiguration/gradle/compileClasspath.lockfile:
```diff
@@ -0,0 +1,5 @@
+# This is a Gradle generated file for dependency locking.
+# Manual edits can break the build and are not advised.
+# This file is expected to be part of source control.
+org:bar:1.3
```
It might be interesting to see transitive dependencies in the example here.

eriwen commented on subprojects/docs/src/docs/userguide/dependencyLocking.adoc:
```diff
@@ -0,0 +1,144 @@
+* Run `./gradlew dependencies --write-locks`.
+This will effectively lock all resolvable configurations that have locking enabled.
+Note that in a multi project setup, `dependencies` only is executed on _one_ project, the root one in this case.
+* Declare a custom task that will resolve all configurations
```
❓ Do we even need to document this 2nd option? Why would I want to do this instead?

ljacomet replied:
As you may need to resolve only specific configurations. For example in the native world, you may not be able to resolve all configurations at once as they might require to be resolved on different platforms.

ljacomet replied:
Added a word to the doc about that example.

eriwen commented on subprojects/docs/src/docs/release/notes.md:
```diff
@@ -40,6 +40,11 @@ The Signing plugin now supports signing all artifacts of a publication, e.g. whe
 
 The [Ivy Publishing Plugin](userguide/publishing_ivy.html) now writes dependency exclude rules defined on a configuration (instead of on an individual dependency) into the generated Ivy module descriptor; the [Maven Publishing Plugin](userguide/publishing_maven.html) now repeats them for each dependency in the generated POM.
 
+### Locking of dynamic dependencies
```
It looks like we have some good content from the intro of the doc we can reuse here, including a short code example.

ljacomet replied:
Done

eriwen commented on subprojects/docs/src/docs/userguide/dependencyLocking.adoc:
```diff
@@ -0,0 +1,144 @@
+
+For this, you have two options:
+
+* Run `./gradlew dependencies --write-locks`.
```
❓ Probably omit the use of Gradle wrapper here.

oesolutions commented on subprojects/docs/src/docs/userguide/dependencyLocking.adoc:
```diff
@@ -0,0 +1,144 @@
+
+=== Locking limitations
+
+It is currently not possible to lock the `classpath` configuration used for build plugins.
```
Are there plans to support this in the future? It's definitely something I think would be useful.
We like to use dynamic dependencies for our plugins in order to have plugin changes automatically applied. While we do our best not to break things within the same major version of a plugin, mistakes happen. If the plugin versions were locked, then the integration of new plugin versions could be controlled with the same infrastructure we use for controlling the automatic integration of the project dependencies.

pioterj commented on subprojects/docs/src/docs/userguide/dependencyLocking.adoc:
```diff
@@ -0,0 +1,144 @@
+=== Nebula locking plugin
+
+This feature is inspired by the https://github.com/nebula-plugins/gradle-dependency-lock-plugin[Nebula Gradle dependency lock plugin].
+It is effectively an adaptation of the features of that plugin that leverage the recent developments around dependency management in Gradle.
```
Let's avoid mentioning recency in the user guide. It will become outdated before we notice.

melix replied:
It seems we have 2 camps there. I told @ljacomet to remove the mention to Nebula, when @eriwen asked him to integrate back. I tend to think like you Piotr that it's too simple to get outdated with an explicit mention. If we care about migration, a blog post or a guide would be better suited I think.

pioterj replied:
This is a valid concern especially given that the core capability will essentially supersede the Nebula plugin. Perhaps the Nebula plugin documentation would be a better place to document the migration and we could add a link here? What do you think @eriwen and @ljacomet?

eriwen replied:
The important thing is that we give those using the Nebula plugins a migration resource. Nebula plugin documentation would be an appropriate place, but I think it's on us to work with @DanielThomas and others to ensure it's gets there if we all agree.

ljacomet replied:
I rephrased this to still have the "inspired by" reference, but then only send you there if you are migrating.

pioterj commented on subprojects/docs/src/docs/userguide/dependencyLocking.adoc:
```diff
@@ -0,0 +1,144 @@
+
+There are differences though between the plugin features and what is integrated into Gradle core.
+
+The main difference is that the scope of locking in Gradle is per configuration, and thus implicitly per project.
```
This doesn't explain the difference. What is the scope of locking in the Nebula plugin?

eriwen replied:
My understanding is that there were some other limitations of the plugin that have now been lifted. I hope we can say something about those in whatever migration text we have.

ljacomet replied:
Removed

ljacomet commented on subprojects/dependency-management/preview-features.adoc:
```diff
@@ -10,7 +10,7 @@ The new metadata format is still under active development. The latest version of
 
 [NOTE]
-Activate this feature in Gradle 4.6+ by setting `org.gradle.gradlemetadata=true` in _gradle.properties_.
+Activate this feature in Gradle 4.6+ by adding `enableFeaturePreview("GRADLE_METADATA")` in _settings.gradle(.kts)_.
```
No, that change is unrelated to dependency locking. It is in a separate commit and makes that document up to date. I missed it when we moved to the `enableFeaturePreview` API.

ljacomet commented on subprojects/docs/src/docs/userguide/dependencyLocking.adoc:
```diff
@@ -0,0 +1,144 @@
+// limitations under the License.
+
+[[dependency_locking]]
+== Dependency locking
```
Fixed

ljacomet commented on subprojects/docs/src/docs/userguide/dependencyLocking.adoc:
```diff
@@ -0,0 +1,144 @@
+[[dependency_locking]]
+== Dependency locking
+
+=== Why dependency locking?
```
Removed

ljacomet commented on subprojects/docs/src/docs/userguide/dependencyLocking.adoc:
```diff
@@ -0,0 +1,144 @@
+
+=== Why dependency locking?
+
+If your project uses dynamic versions anywhere, the resolved set of dependencies varies through time.
```
Applied

ljacomet commented on subprojects/docs/src/docs/userguide/dependencyLocking.adoc:
```diff
@@ -0,0 +1,144 @@
+** A version mismatch or missing resolved module causes a build failure
+* Resolution result must not contain extra dependencies compared to the lock file content
+
+=== Generating and updating lockfiles
```
Moved

ljacomet commented on subprojects/docs/src/docs/userguide/dependencyLocking.adoc:
```diff
@@ -0,0 +1,144 @@
+</sample>
+++++
+
+=== Controlled lock file update
```
Changed

ljacomet commented on subprojects/docs/src/docs/userguide/dependencyLocking.adoc:
```diff
@@ -0,0 +1,144 @@
+Wildcards, indicated with `*`, can be used in the group or module name. They can be the only character or appear at the end of the group or module respectively.
+The following wildcard notations are valid:
+
+* `org*:foo`
```
Added

ljacomet replied:
And used well known examples

ljacomet commented on subprojects/docs/src/docs/userguide/dependencyLocking.adoc:
```diff
@@ -0,0 +1,144 @@
+Note however that the resolution may cause other module versions to update, as dictated by the Gradle resolution rules.
+
+----
+❯ ./gradlew classes --update-locks org:foo,org:bar
```
Updated with real world examples

ljacomet commented on subprojects/docs/src/docs/userguide/dependencyLocking.adoc:
```diff
@@ -0,0 +1,144 @@
+* `*:bar`
+* `org*:bar*`
+
+=== Deleting a lockfile
```
Changed

ljacomet commented on subprojects/docs/src/docs/userguide/troubleshootingDependencyResolution.adoc:
```diff
@@ -86,9 +86,17 @@ The `--refresh-dependencies` option tells Gradle to ignore all cached entries fo
+In the best case the CI build catches the problem and someone needs to investigate.
+In the worst case, the problem makes it to production unnoticed.
+
+Gradle now offers <<dependency_locking,dependency locking>> to solve this problem.
```
Removed

ljacomet commented on subprojects/docs/src/docs/release/notes.md:
```diff
@@ -40,6 +40,11 @@ The Signing plugin now supports signing all artifacts of a publication, e.g. whe
 
+### Locking of dynamic dependencies
+
+Gradle core now provides a mechanism for [locking dynamic versions](userguide/dependency_locking.html).
```
Added

ljacomet commented:
@melix @eriwen @pioterj I have addressed all comments, aside from the section ordering in the `userGuideHtml.xsl`

Let me know if there are further comments.

pioterj commented on subprojects/docs/src/docs/release/notes.md:
```diff
@@ -6,6 +6,25 @@ Here are the new features introduced in this Gradle release.
+
+    gradle test --write-locks
+
+From this point onwards, all configurations that have a lock state wil lfail to resolve if changes are made to their dependencies.
```
There's a typo in "wil lfail".

pioterj commented on subprojects/docs/src/docs/userguide/dependencyLocking.adoc:
```diff
@@ -0,0 +1,152 @@
+
+This feature is inspired by the https://github.com/nebula-plugins/gradle-dependency-lock-plugin[Nebula Gradle dependency lock plugin].
+
+If you were using that plugin and want to migrate, please see the Nebula documentation.
```
I'd remove the word "please". It is generally considered needless politeness in technical writing. I learned that from professional technical writers I worked with a while ago. See also http://www.drdobbs.com/architecture-and-design/10-things-tech-writers-want-developers-t/240165685 and https://developers.google.com/style/tone#politeness-and-use-of-please.

eriwen commented on subprojects/docs/src/docs/release/notes.md:
```diff
@@ -6,6 +6,25 @@ Here are the new features introduced in this Gradle release.
 IMPORTANT: if this is a patch release, ensure that a prominent link is included in the foreword to all releases of the same minor stream.
 Add-->
 
+### Locking of dynamic dependencies
```
Very nice 👍
